source/beamline/SGM/energy/SGMEnergyControl.cpp

```diff
@@ -0,0 +1,363 @@
+
+bool SGMEnergyControl::isUndulatorTracking() const
+{
+    if(energyPositionController_) {
+
+        return energyPositionController_->isUndulatorTracking();
+    }
+
+    return false;
+}
+
+SGMEnergyPosition::GratingTranslationOptimizationMode SGMEnergyControl::gratingTranslationOptimizationMode() const
+{
+    if(energyPositionController_) {
+
+        return energyPositionController_->gratingTranslationOptimizationMode();
+    }
+
+    return SGMEnergyPosition::ManualMode;
+}
+
+bool SGMEnergyControl::isExitSlitPositionTracking() const
+{
+    if(energyPositionController_) {
+
+        return energyPositionController_->isExitSlitPositionTracking();
+    }
+
+    return false;
+}
+
+AMControl *SGMEnergyControl::gratingAngleControl() const
+{
+    return gratingAngleControl_;
+}
+
+AMControl *SGMEnergyControl::gratingTranslationControl() const
+{
+    return gratingTranslationControl_;
+}
+
+AMControl *SGMEnergyControl::undulatorPositionControl() const
+{
+    return undulatorPositionControl_;
+}
+
+AMControl *SGMEnergyControl::exitSlitPositionControl() const
+{
+    return exitSlitPositionControl_;
+}
+
+void SGMEnergyControl::setUndulatorHarmonic(SGMUndulatorSupport::UndulatorHarmonic undulatorHarmonic)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorHarmonic(undulatorHarmonic);
+    }
+}
+
+void SGMEnergyControl::setUndulatorOffset(double undulatorOffset)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorOffset(undulatorOffset);
+    }
+}
+
+void SGMEnergyControl::setUndulatorTracking(bool isTracking)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setUndulatorTracking(isTracking);
+    }
+}
+
+void SGMEnergyControl::setGratingTranslationOptimizationMode(SGMEnergyPosition::GratingTranslationOptimizationMode gratingTranslationOptimizationMode)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setGratingTranslationOptimizationMode(gratingTranslationOptimizationMode);
+    }
+}
+
+void SGMEnergyControl::setExitSlitPositionTracking(bool isTracking)
+{
+    if(energyPositionController_) {
+
+        energyPositionController_->setExitSlitPositionTracking(isTracking);
+    }
+}
+
+void SGMEnergyControl::updateConnected()
+{
+    bool nowConnected = gratingAngleControl_->isConnected() &&
+            gratingTranslationControl_->isConnected() &&
+            undulatorPositionControl_->isConnected() &&
+            exitSlitPositionControl_->isConnected();
+
+    if(nowConnected && !energyPositionController_) {
+
+        initializeEnergyPositionController();
+    }
+
+    setConnected(nowConnected);
+}
+
+void SGMEnergyControl::updateMoving()
+{
+    setIsMoving(gratingAngleControl_->isMoving() ||
+                gratingTranslationControl_->isMoving() ||
+                undulatorPositionControl_->isMoving() ||
+                exitSlitPositionControl_->isMoving());
+}
+
+void SGMEnergyControl::updateValue()
```

Comment:
I'm guessing that updateValue() doesn't work for you in this context, if it's not implemented here. Is it currently pure virtual in AMPseudoMotorControl? Maybe that should change.